Implement xdebug.discover_client_host, fix bug with WSL2 docker-inside getting correct host.docker.internal, fixes #4803 (#4852)
ddev/ddev: docs/content/users/debugging-profiling/step-debugging.md

* `ddev ssh` and try the `telnet host.docker.internal 9003` again. It should connect. If not, maybe PhpStorm is not listening, or not configured to listen on port 9003?
* Check to make sure that Xdebug is enabled. You can use `php -i | grep -i xdebug` inside the container, or use any other technique you want that gives the output of `phpinfo()`, including Drupal’s `admin/reports/status/php`. You should see `with Xdebug v3` and `php -i | grep xdebug.mode` should give you `xdebug.mode => debug,develop => debug,develop"`.
* Set a breakpoint in the first relevant line of your `index.php` and then visit the site in a browser. It should stop at that first line.

> * If you're using a flavor of IDE that connects directly into the web container like VS Code Language Server, you may want to use the [global configuration xdebug_ide_location](../configuration/config.md#xdebugidelocation) to explain to DDEV the situation. For example, `ddev config global --xdebug-ide-location=container`, which tells the PHP/Xdebug to connect directly to the listener inside the container.

Reviewer: I’d wrap the config setting in backticks and refer to it as a “setting” in an effort to use that language consistently and help subtly reinforce how you use DDEV.

```suggestion
* If you're using a flavor of IDE that connects directly into the web container like VS Code Language Server, you may want to use the [global `xdebug_ide_location` setting](../configuration/config.md#xdebugidelocation) to explain to DDEV the situation. For example, `ddev config global --xdebug-ide-location=container`, which tells the PHP/Xdebug to connect directly to the listener inside the container.
```

### WSL2 Xdebug Troubleshooting

WSL2 is a complicated environment for Xdebug, especially if you're running your IDE on the Windows side, as most people do.

> * With PhpStorm, consider using the "Remote Development" feature to connect to WSL. That runs an acrtual PhpStorm instance on WSL2 so the network complications are not as great.

Reviewer: Fix typo, rephrase to describe the affirmative.

```suggestion
* With PhpStorm, consider using the "Remote Development" feature to connect to WSL. That runs an actual PhpStorm instance on WSL2 to reduce networking complexity.
```

> * When using an IDE inside WSL2 (for example, when running PhpStorm or Visual Studio Code inside WSL2, or when using PhpStorm's "Remote Development" feature) you may need to tell Xdebug about this situation with `ddev config global --xdebug-ide-location=wsl2`, see [xdebug_ide_location](../configuration/config.md#xdebugidelocation).

Reviewer: Make this slightly easier to read, mention the _setting_ up front and in backticks, try to be more specific about what the setting is doing (which I hope I got right):

```suggestion
* When using an IDE inside WSL2—like you would running PhpStorm or Visual Studio Code inside WSL2, or using PhpStorm's "Remote Development" feature—you may need to use the [`xdebug_ide_location`](../configuration/config.md#xdebugidelocation) setting to tell Xdebug to expect the IDE under WSL2. You can do this by running `ddev config global --xdebug-ide-location=wsl2`.
```

* `export DDEV_DEBUG=true && ddev start` will show you how DDEV is calculating the `host.docker.internal` IP address to be used when contacting the IDE, which may give a hint about problems you might discover in the general troubleshooting discussed above, when trying to connect to the listening IDE.

> * If you're using the IDE on the Windows side, the WSL2 file /etc/resolv.conf is supposed to be the authoritative way to figure out where the IDE is in the Windows networking scheme. That value should be the same as `host.docker.internal`, so `ddev exec ping -c 1 host.docker.internal` will show you what's actually being used. If your IDE is actually at a different address, you can tell DDEV to override the discovered value for `host.docker.internal` with `ddev config global --xdebug-ide-location=<some_ip_address>`.

Reviewer: Backticks, improve flow, run commands:

```suggestion
* If you're using the IDE on the Windows side, WSL2’s  `/etc/resolv.conf` file is supposed to be the authoritative way to figure out where the IDE is in the Windows networking scheme. That value should be the same as `host.docker.internal`, so running `ddev exec ping -c 1 host.docker.internal` will show you what's actually being used. If your IDE is actually at a different address, you can tell DDEV to override the discovered value for `host.docker.internal` by running `ddev config global --xdebug-ide-location=<some_ip_address>`.
```
